### PR TITLE
Fix lexical scoping between scripts

### PR DESCRIPTION
**Pull request by dbatyai**
This change fixes the handling of lexical blocks when executing multiple
scripts, and also fixes a few issues with module environments.

After this change, all script files will run in the same context and
will have access to lexically scoped global variables of previous
scripts, and module environments will no longer have a bound global 'this'
value.

The REPL implementation in main-unix is also fixed to correctly handle
lexically scoped variables.

### CODE REVIEW

File: jerry-core/parser/js/js-scanner-util.c [hunk 1]
```diff
@@ -1302,7 +1299,8 @@ scanner_detect_invalid_var (parser_context_t *context_p, /**< context */
     }
   }
 
-  if ((context_p->status_flags & PARSER_IS_EVAL)
+  if (((context_p->status_flags & PARSER_IS_EVAL)
+       || literal_pool_p->prev_p == NULL)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-scanner-util.c]: This can be kept in one line.

**Comment by dbatyai** [comment]
In the standard parsing and execution cannot be done separately, each specified scipt should be parsed exactly once, and if no error occured, should be executed exactly once immediatly after, so normally this shouldn't be a problem.

However, since in our case parsing and exection can be done separately, we'll either have to do runtime checks for previous `let` declarations if we would like to allow executing a parsed script multiple times or in different order, or we''ll have to make sure that evaluation is done according to the standard.

File: jerry-core/parser/js/js-scanner-util.c [hunk 2]
```diff
@@ -1738,6 +1731,96 @@ scanner_create_unused_literal (parser_context_t *context_p, /**< context */
...
+  JERRY_ASSERT (info_p->type == SCANNER_TYPE_FUNCTION);
+
+  literal.char_p = info_p->source_p - 1;
+
+  while (next_data_p[0] != SCANNER_STREAM_TYPE_END)
+  {
+    uint32_t type = next_data_p[0] & SCANNER_STREAM_TYPE_MASK;
+    const uint8_t *data_p = next_data_p;
+
+    if (type == SCANNER_STREAM_TYPE_HOLE)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-scanner-util.c]: I think this type is not allowed in global scope. I would add an assert instead.

File: jerry-core/parser/js/js-scanner-util.c [hunk 3]
```diff
@@ -1738,6 +1731,96 @@ scanner_create_unused_literal (parser_context_t *context_p, /**< context */
...
+    {
+      next_data_p++;
+      continue;
+    }
+
+    JERRY_ASSERT (type != SCANNER_STREAM_TYPE_HOLE
```
Comment (zherczeg) [inline jerry-core/parser/js/js-scanner-util.c]: The assert is here, just delete the lines above.

**Pull request merged**